src/layers/retry.rs

```diff
@@ -639,6 +651,137 @@ impl<R: output::BlockingRead> output::BlockingRead for RetryReader<R> {
     }
 }
 
+/// the retriable pager implementation
+#[derive(Debug)]
+pub struct RetryPager<P> {
+    // object pager
+    inner: P,
+    // used for logging
+    path: String,
+    // backoff policy of the pager
+    policy: ExponentialBuilder,
+    // the current backoff chain
+    // Note:
+    // each polling of pages has its own backoff chain
+    // once the poll is success, the backoff chain will be reset.
+    current_backoff: Option<ExponentialBackoff>,
+    // backoff time to sleep
+    sleep: Option<Duration>,
+}
+
+impl<P> RetryPager<P> {
+    fn new(inner: P, path: &str, policy: ExponentialBuilder) -> Self {
+        Self {
+            inner,
+            path: path.to_string(),
+            policy,
+            current_backoff: None,
+            sleep: None,
+        }
+    }
+
+    fn reset_backoff(&mut self) {
+        self.current_backoff = None;
+        self.sleep = None;
+    }
+}
+
+#[async_trait]
+impl<P: ObjectPage> ObjectPage for RetryPager<P> {
+    async fn next_page(&mut self) -> Result<Option<Vec<ObjectEntry>>> {
```

Comment:
How about using the `Retryable` trait directly? I guess we don't need to implement the retry logic by hand.

Reply:
Applied on synchronous API.
However, for asynchronous API, it's kind of hard to use this trait. Using `{|| self.inner.next_page()}` will make Rustc complain about `self` escaping this closure, I have no idea how to solve it.